clang/lib/CodeGen/CGHLSLRuntime.cpp

```diff
@@ -588,7 +588,7 @@ llvm::Function *CGHLSLRuntime::createResourceBindingInitFn() {
       auto *Slot = llvm::ConstantInt::get(CGM.IntTy, RBA->getSlotNumber());
       // FIXME: resource arrays are not yet implemented
       auto *Range = llvm::ConstantInt::get(CGM.IntTy, 1);
-      auto *Index = llvm::ConstantInt::get(CGM.IntTy, 0);
+      auto *Index = llvm::ConstantInt::get(CGM.IntTy, RBA->getSlotNumber());
```

Comment:
This is not the way I implemented this for SPIR-V. This assumes the `Slot` and `Index` as interwined, which they are in DXIL, but not in SPIR-V.

In SPIR-V, the `Slot` is the binding number, and all elements of the resource array are accessed through the same binding number. The element does not get their own binding number. The binding and index are independent.

Would it be possible to keep this as is, and have the DXIL backend add the slot number to the index?

Reply:
It might be better to just diverge this by the target. That way the intermediate IR reflects the index correctly.

Reply:
Diverging is fine. We will have to diverge when identifying the slots anyway. Because we keep samplers and textures in the same "space", we will need a more complicated way of determining the right space and slot. Plus we will eventually at the `vk::binding` attribute which to be another consideration.

Reply:
If we opt to diverge this by target we should probably not use the matching intrinsic names between dxil and spirv in this case, as that could end up confusing. It might be best to defer that until we're dealing with textures or the binding attribute so we can make sure we're making the right decisions on how to split it up.

I *think* it would be fine to just do this indexing in the backend, since any index that's less than the slot should be invalid anyway. I'll try that out and report back.

Reply:
I've changed this to do the translation in the backend.